src/samVisualize/rendering/forceDirectedGraph.ts

```diff
@@ -0,0 +1,668 @@
+     */
+    private iconPaths: { [resourceType: string]: string }
+
+    private vscodeApi: WebviewApi<unknown> | undefined
+
+    // The overarching g container containing nodes & links
+    // Having this separate container element allows for zooms and pans
+    private g: d3.Selection<SVGGElement, unknown, HTMLElement | null, unknown>
+
+    // The d3.Simulation which runs the physics and interactivity of the force directed graph
+    public simulation: d3.Simulation<NodeDatum, LinkDatum>
+
+    // Since the link construction uses the id of the arrowheads,
+    // it's declared and passed around to avoid mismatching string literals.
+    public readonly arrowheadID = 'arrowhead'
+
+    private links: d3.Selection<SVGGElement, LinkDatum, SVGGElement, unknown> | undefined
+    private nodes: d3.Selection<SVGGElement, NodeDatum, SVGGElement, unknown> | undefined
+
+    // Holds the div containing the filter radio buttons
+    // Visibility is toggled during error renders
+    private filterButtonsDiv: d3.Selection<HTMLDivElement, unknown, HTMLElement | null, unknown>
+
+    constructor(
+        completeGraphObject: GraphObject,
+        primaryResourceList: string[],
+        iconPaths: { [resourceType: string]: string },
+        doc?: Document
+    ) {
+        this.completeGraphObject = completeGraphObject
+        this.iconPaths = iconPaths
+
+        try {
+            this.vscodeApi = acquireVsCodeApi()
+        } catch {
+            // Test environment
+            this.vscodeApi = undefined
+        }
+
+        let svgWidth, svgHeight
+        try {
+            svgWidth = document.querySelector('body')!.clientWidth
+            svgHeight = document.querySelector('body')!.clientHeight
+        } catch {
+            // Test environment
+            svgWidth = 500
+            svgHeight = 500
+        }
+
+        const svg = this.constructSVG(svgWidth, svgHeight, 'svg', doc)
+        this.g = this.appendGContainerElement(svg)
+
+        this.defineArrowHead(svg)
+
+        // Define forces
+        this.simulation = d3
+            .forceSimulation<NodeDatum, LinkDatum>()
+            .force(
+                'link',
+                d3
+                    .forceLink<NodeDatum, LinkDatum>()
+                    .id((d: NodeDatum) => d.name)
+                    .distance(PhysicsConfig.linkForceDistance)
+            )
+            .force('charge', d3.forceManyBody().strength(PhysicsConfig.nodeForce))
+            .force('center', d3.forceCenter(svgWidth / 2, svgHeight / 2))
+            .force('forceX', d3.forceX(svgWidth / 2).strength(PhysicsConfig.centeringForceStrength))
+            .force('forceY', d3.forceY(svgHeight / 2).strength(PhysicsConfig.centeringForceStrength))
+
+        this.simulation.alphaTarget(PhysicsConfig.alphaTarget)
+        this.simulation.alphaDecay(PhysicsConfig.alphaDecay)
+        // Don't stop ticking
+        this.simulation.alphaMin(-1)
+
+        try {
```

Comment:
I'm not familiar with testing web views but using these try blocks with empty catches seems kind of odd. Is there a better way to do this?

Reply:
In the PR desc I mention that they're only there to avoid crashes during testing. I tried to refactor to avoid calling untestable code while keeping the rendering process reasonable, but the try catch seemed most pragmatic. I found testing d3.js somewhat difficult in general. Maybe environment variables instead?

Reply:
Ah yeah I started reading the code before the PR desc, my bad.

Reply:
Update: Replaced with clear if statements, depending on the presence of a test Document (upon which D3 draws on during unit tests